File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    nums.sort((a, b) => a - b)
    
    const N = nums.length
    
    const answer = []
    
    for (let i = 0; i < N; i++) {
        if (i > 0 && nums[i - 1] === nums[i]) {
            continue
        }
        
        let left = i + 1
        let right = N - 1
        while (left < right) {
            const threeSum = nums[i] + nums[left] + nums[right]
            
            if (threeSum < 0) {
                left++
            } else if (threeSum > 0) {
                right--
            } else {
                answer.push([nums[i], nums[left], nums[right]])
                left++
                while (nums[left - 1] === nums[left] && left < right) {
                    left++
                }
            }
        }
    }
    
    return answer
};

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> answer = new LinkedList<>();
        
        Arrays.sort(nums);
        for (int i = 0; i < nums.length; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            int left = i + 1;
            int right = nums.length - 1;
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum < 0) {
                    left++;
                } else if (sum > 0) {
                    right--;
                } else {
                    List<Integer> list = new LinkedList<>();
                    list.add(nums[i]);
                    list.add(nums[left]);
                    list.add(nums[right]);
                    answer.add(list);
                    left++;
                    while (nums[left] == nums[left - 1] && left < right) {
                        left++;
                    }
                }
            }
        }
        
        return answer;
    }
}